[AIRFLOW-3611] Breeze (#4932)
mistercrunch/Airflow: BREEZE.rst

* `Keeping images up-to-date <#keeping-images-up-to-date>`_
    - `Updating dependencies <#updating-dependencies>`_
    - `Pulling the images <#pulling-the-images>`_
* `Airflow Breeze flags <#airflow-breeze-flags>`_

Airflow Breeze
==============

Airflow Breeze is an easy-to-use integration test environment managed via
`Docker Compose <https://docs.docker.com/compose/>`_ .
The environment is easy to use locally and it is also used by Airflow's CI Travis tests.

It's called **Airflow Breeze** as in "It's a *Breeze* to develop Airflow"

The advantages and disadvantages of using the environment vs. other ways of testing Airflow
are described in `CONTRIBUTING.md <CONTRIBUTING.md#integration-test-development-environment>`_.

Here is the short 10 minute video about Airflow Breeze

.. image:: http://img.youtube.com/vi/ffKFHV6f3PQ/0.jpg
   :width: 480px
   :height: 360px
   :scale: 100 %
   :alt: Airflow Breeze Simplified Development Workflow
   :align: center
   :target: http://www.youtube.com/watch?v=ffKFHV6f3PQ

Installation
============

Prerequisites for the installation:

*
  If you are on MacOS you need gnu getopt to get Airflow Breeze running. Typically

>   uou need to run ``brew install gnu-getopt`` and then follow instructions (you need

Reviewer: ```suggestion
  you need to run ``brew install gnu-getopt`` and then follow instructions (you need
```

Author: Done.

  to link the gnu getopt version to become first on the PATH).

*
  Latest stable Docker Community Edition installed and on the PATH. It should be
  configured to be able to run ``docker`` commands directly and not only via root user

  * your user should be in ``docker`` group.
    See `Docker installation guide <https://docs.docker.com/install/>`_

*
  Latest stable Docker Compose installed and on the PATH. It should be
  configured to be able to run ``docker-compose`` command.
  See `Docker compose installation guide <https://docs.docker.com/compose/install/>`_

*

>   in case of MacOS you need ``gstat`` installed via brew or port. The stat command from MacOS

Reviewer: Can we group this one with the 1st point about installing `gnu-getopt`

Author: Done.

You can add the checked out airflow repository to your PATH to run breeze
without the ./ and from any directory if you have only one airflow directory checked out.

See `Airflow Breeze flags <#airflow-breeze-flags>`_ for details.

First time you run `./breeze <./breeze>`_ script, it will pull and build local version of docker images.
It will pull latest Airflow CI images from `Apache Airflow DockerHub <https://hub.docker.com/r/apache/airflow>`_
and use them to build your local docker images. It will use latest sources from your source code.
Further on ``breeze`` uses md5sum calculation and Docker caching mechanisms to only rebuild what is needed.
Airflow Breeze will detect if Docker images need to be rebuilt and ask you for confirmation.

Setting up autocomplete
=======================

The ``breeze`` command comes with built-in bash/zsh autocomplete. When you start typing
`./breeze <./breeze>`_ command you can use <TAB> to show all the available switches
nd to get autocompletion on typical values of parameters that you can use.

You can setup auto-complete automatically by running this command (-a is shortcut for --setup-autocomplete):

.. code-block:: bash

>    ./breeze --setup-autocomplete

Reviewer: Awesome

Author: And it also works for .zsh  (limited to options for now)

restore default settings.

Relevant sources of airflow are mounted inside the ``airflow-testing`` container that you enter,
which means that you can continue editing your changes in the host in your favourite IDE and have them
visible in docker immediately and ready to test without rebuilding images. This can be disabled by specifying
``--skip-mounting-source-volume`` flag when running breeze, in which case you will have sources
embedded in the container - and changes to those sources will not be persistent.

Once you enter the environment you are dropped into bash shell and you can run tests immediately.

Running tests in Airflow Breeze
-------------------------------

Once you enter Airflow Breeze environment you should be able to simply run
`run-tests` at will. Note that if you want to pass extra parameters to nose
you should do it after '--'

For example, in order to just execute the "core" unit tests, run the following:

.. code-block:: bash

   run-tests tests.core:CoreTest -- -s --logging-level=DEBUG

or a single test method:

.. code-block:: bash

   run-tests tests.core:CoreTest.test_check_operators -- -s --logging-level=DEBUG

> The tests will run 'airflow db reset' and 'airflow db init' the first time you

Reviewer: As they are commands:

```suggestion
The tests will run ``airflow db reset`` and ``airflow db init`` the first time you
```

Author: Right. done

run tests in running container, so you can count on database being initialized.

All subsequent test executions within the same container will run without database
initialisation.

You can also optionally add --with-db-init flag if you want to re-initialize
the database.

.. code-block:: bash

   run-tests --with-db-init tests.core:CoreTest.test_check_operators -- -s --logging-level=DEBUG

Debugging with ipdb
-------------------

You can debug any code you run in the container using ``ipdb`` debugger if you prefer console debugging.
It is as easy as copy&pasting this line into your code:

.. code-block:: python

   import ipdb; ipdb.set_trace()

Once you hit the line you will be dropped into interactive ipdb  debugger where you have colors
and auto-completion to guide your debugging. This works from the console where you started your program.

> Note that in case of `nosetest` you need to provide `--nocapture` flag to avoid nosetests capturing the stdout

Reviewer: ```suggestion
Note that in case of ``nosetest`` you need to provide ``--nocapture`` flag to avoid nosetests capturing the stdout
```

Author: Resolved

      webserver_config.py - webserver configuration generated when running airflow in the container

Note that when run in your local environment ``/root/airflow/logs`` folder is actually mounted from your
``logs`` directory in airflow sources, so all logs created in the container are automatically visible in the host
as well. Every time you enter the container the logs directory is cleaned so that logs do not accumulate.

Port forwarding
---------------

When you run Airflow Breeze, the following ports are automatically forwarded:

* 28080 -> forwarded to airflow webserver -> airflow-testing:8080
* 25433 -> forwarded to postgres database -> postgres:5432
* 23306 -> forwarded to mysql database  -> mysql:3306

You can connect to those ports/databases using:

* Webserver: (http://127.0.0.1:28080)[http://127.0.0.1:28080]
* Postgres: ``jdbc:postgresql://127.0.0.1:25433/airflow?user=postgres&password=airflow``
* Mysql: ``jdbc:mysql://localhost:23306/airflow?user=root``

Note that you need to start the webserver manually with ``airflow webserver`` command if you want to connect
to the webserver (you can use ``tmux`` to multiply terminals).

> For databases you need to run ``airflow resetdb`` at least once after you started Airflow Breeze to get

Reviewer: ```suggestion
For databases you need to run ``airflow db reset`` at least once after you started Airflow Breeze to get
```

Author: Done.

You can create the virtualenv using ``virtualenvwrapper`` - that will allow you to easily switch between
virtualenvs using workon command and mange your virtual environments more easily.

Typically creating the environment can be done by:

.. code-block:: bash

  mkvirtualenv <ENV_NAME> --python=python<VERSION>

After the virtualenv is created, you must initialize it. Simply enter the environment
(using workon) and once you are in it run:

.. code-block:: bash

  ./breeze --initialize-local-virtualenv

Once initialization is done, you should select the virtualenv you initialized as the project's default
virtualenv in your IDE.

Running unit tests via IDE
--------------------------

After setting it up - you can use the usual "Run Test" option of the IDE and have all the
autocomplete and documentation support from IDE as well as you can debug and click-through
the sources of Airflow - which is very helpful during development. Usually you also can run most
of the unit tests (those that do not require prerequisites) directly from the IDE:

Running unit tests from IDE is as simple as:

> .. raw:: html

Reviewer: Just curious why were are using raw html and not the `image` tag

Author: It was automatically converted from an earlier markdonw version and I used raw there to have center alignment (no such feature in markdown). I have not noticed it used RAW :). Fixed now..

   <p align="center">
     <img src="images/running_unittests.png" alt="Running unit tests"/>
   </p>

Some of the core tests use dags defined in ``tests/dags`` folder - those tests should have
``AIRFLOW__CORE__UNIT_TEST_MODE`` set to True. You can set it up in your test configuration:

.. raw:: html

   <p align="center">
     <img src="images/airflow_unit_test_mode.png" alt="Airflow Unit test mode"/>
   </p>

You cannot run all the tests this way - only unit tests that do not require external dependencies
such as postgres/mysql/hadoop etc. You should use
`Running tests in Airflow Breeze <#running-tests-in-airflow-breeze>`_ in order to run those tests. You can
still use your IDE to debug those tests as explained in the next chapter.

Debugging Airflow Breeze Tests in IDE
-------------------------------------

> When you run example DAGs - even if you run them using UnitTests from within IDE, they are run in a separate

Reviewer: ```suggestion
When you run example DAGs, even if you run them using UnitTests from within IDE, they are run in a separate
```

Author: Fixed.

.. raw:: html

   <p align="center">
     <img src="images/setup_remote_debugging.png" alt="Setup remote debugging"/>
   </p>

Not that if you are on ``MacOS`` you have to use the real IP address of your host rather than default
localhost because on MacOS container runs in a virtual machine with different IP address.

You also have to remember about configuring source code mapping in remote debugging configuration to map
your local sources into the ``/opt/airflow`` location of the sources within the container.

.. raw:: html

   <p align="center">
     <img src="images/source_code_mapping_ide.png" alt="Source code mapping"/>
   </p>

Running commands via Airflow Breeze
===================================

Running static code checks
--------------------------

If you wish to run static code checks inside Docker environment you can do it via
``-S``, ``--static-check`` flags or ``-F``, ``--static-check-all-files``. The former will run appropriate
checks only for files changed and staged locally, the latter will run it on all files. It can take a lot of
time to run check for all files in case of pylint on MacOS due to slow filesystem for Mac OS Docker.

> You can add arguments you should pass them after -- as extra arguments.

Reviewer: ```suggestion
You can add arguments you should pass them after ``--`` as extra arguments.
```

The documentation is build using ``-O``, ``--build-docs`` command:

.. code-block:: bash

     ./breeze --build-docs

Results of the build can be found in ``docs/_build`` folder. Often errors during documentation generation
come from the docstrings of auto-api generated classes. During the docs building auto-api generated
files are stored in ``docs/_api`` folder - so that in case of problems with documentation you can
find where the problems with documentation originated from.

Running tests
-------------

If you wish to run tests only and not drop into shell, you can run them by providing
-t, --test-target flag. You can add extra nosetest flags after -- in the commandline.

.. code-block:: bash

     ./breeze --test-target tests/hooks/test_druid_hook.py -- --logging-level=DEBUG

You can run the whole test suite with special '.' test target:

.. code-block:: bash

    ./breeze --test-target .

You can also specify individual tests or group of tests:

.. code-block:: bash

    ./breeze --test-target tests.core:CoreTest

Running commands inside Docker
------------------------------

If you wish to run other commands/executables inside of Docker environment you can do it via

> -x, --execute-command flag. Note that if you want to add arguments you should specify them

Reviewer: ```suggestion
``-x``, ``--execute-command`` flag. Note that if you want to add arguments you should specify them
```

Author: Done

> together withe the command surrounded with " or ' or pass them after -- as extra arguments.

Reviewer: ```suggestion
together with the command surrounded with " or ' or pass them after -- as extra arguments.
```

Author: Done.

.. code-block:: bash

     ./breeze --execute-command "ls -la"

.. code-block:: bash

     ./breeze --execute-command ls -- --la

Running Docker Compose commands
-------------------------------

If you wish to run docker-compose command (such as help/pull etc. ) you can do it via

> -d, --docker-compose flag. Note that if you want to add extra arguments you should specify them

Reviewer: This is just my preference, not mandatory. You can ignore this, I don't have a strong opinion :)

```suggestion
``-d``, ``--docker-compose`` flag. Note that if you want to add extra arguments you should specify them
```

Author: It makes perfect sense. I just missed it here, also have the same preference. Fixed.

will invalidate cache for most of the image and it will take long time to rebuild the image by breeze.
Therefore it is a recommended practice to add new dependencies closer to the bottom of
Dockerfile during development (to get the new dependencies incrementally added) and only move them to the
top when you are close to finalise the PR and merge the change. It's OK for development time to add separate
``apt-get install`` commands similar to those that are already there (but remember to move newly added
dependencies to the appropriate ``apt-get install`` command which is already in the Dockerfile.

Pulling the images
------------------

Sometimes the image on DockerHub is rebuilt from the scratch. This happens for example when there is a
security update of the python version that all the images are based on. In this case it is much faster to
pull latest images rather than rebuild them from the scratch. Airflow Breeze will detect such case and
will ask you to confirm to pull and build the image and if you answer OK, it will pull and build the image.
You might also provide ``--force-pull-images`` flag to force pull latest images from DockerHub.

> Airflow Breeze flags

Reviewer: I don't think we should list all the commands in the `.rst` file. The main reason I think that is because if we change a flag in the original script or modify more flags it is easy to forget to update that in the doc.

And a user can simply get this by `breeze --help` and will probably make this file small & concise :)

Author: It's generally nice if you can see all the commands in the web docs. I think this is something I always look for in this kind of documentation (there are people who prefer web docs than command line --help in general). Also it makes it google-searchable and web-viewable, which is rather nice if you are looking for some answers and you have no tools handy (like when you are answering someone's comments on slack and you only have your phone with you),

For example all gcloud commands have something like that: https://cloud.google.com/sdk/gcloud/reference/

I wanted to make it automated in the near future (like doctoc in .md files) so it does not get out-of-sync - this should be rather easy to do but it's already a huge change and I would love to get it out :).

====================

> Those are the current flags of the `./breeze <./breeze>`_ script

Reviewer: ```suggestion
These are the current flags of the `./breeze <./breeze>`_ script
```

Author: Fixed.

* `Using your host IDE <#using-your-host-ide>`_
    - `Configuring local virtualenv <#configuring-local-virtualenv>`_
    - `Running unit tests via IDE <#running-unit-tests-via-ide>`_
    - `Debugging Airflow Breeze Tests in IDE <#debugging-airflow-breeze-tests-in-ide>`_
* `Running commands via Airflow Breeze <#running-commands-via-airflow-breeze>`_
    - `Running static code checks <#running-static-code-checks>`_
    - `Building the documentation <#building-the-documentation>`_
    - `Running tests <#running-tests>`_
    - `Running commands inside Docker <#running-commands-inside-docker>`_
    - `Running Docker Compose commands <#running-docker-compose-commands>`_
    - `Convenience scripts <#convenience-scripts>`_
* `Keeping images up-to-date <#keeping-images-up-to-date>`_
    - `Updating dependencies <#updating-dependencies>`_
    - `Pulling the images <#pulling-the-images>`_
* `Airflow Breeze flags <#airflow-breeze-flags>`_

Airflow Breeze
==============

Airflow Breeze is an easy-to-use integration test environment managed via
`Docker Compose <https://docs.docker.com/compose/>`_ .
The environment is easy to use locally and it is also used by Airflow's CI Travis tests.

> It's called **Airflow Breeze** as in "It's a *Breeze* to develop Airflow"

Reviewer 2: 🤦‍♂ *sigh*... I'll allow it.

Reviewer 3: 😂😂😂😂😂

* Webserver: (http://127.0.0.1:28080)[http://127.0.0.1:28080]
* Postgres: ``jdbc:postgresql://127.0.0.1:25433/airflow?user=postgres&password=airflow``
* Mysql: ``jdbc:mysql://localhost:23306/airflow?user=root``

Note that you need to start the webserver manually with ``airflow webserver`` command if you want to connect
to the webserver (you can use ``tmux`` to multiply terminals).

For databases you need to run ``airflow resetdb`` at least once after you started Airflow Breeze to get
the database/tables created. You can connect to databases with IDE or any other Database client:

.. raw:: html

   <p align="center">
     <img src="images/database_view.png" alt="Database view"/>
   </p>

You can change host port numbers used by setting appropriate environment variables:

* WEBSERVER_HOST_PORT
* POSTGRES_HOST_PORT
* MYSQL_HOST_PORT

When you set those variables, next time when you enter the environment the new ports should be in effect.

Using your host IDE
===================

Configuring local virtualenv
----------------------------

In order to use your host IDE (for example IntelliJ's PyCharm/Idea) you need to have virtual environments

> setup. Ideally you should have virtualenvs for all python versions that Airflow supports (2.7, 3.5, 3.6).

Reviewer 2: Do we still want to suppor 2.7 here?

Author: Right :). Should be 3.7 in this branch :). I just enabled 3.7 master When I cherry-pick to v1-10-test it will be 2.7/3.5/3.6 there

Author: Resolved
